File: beagle-declarative/src/main/kotlin/br/com/zup/beagle/widget/layout/FlexSingleWidget.kt
package br.com.zup.beagle.widget.layout

import br.com.zup.beagle.widget.core.Appearance
import br.com.zup.beagle.widget.core.Flex
import br.com.zup.beagle.widget.core.AppearanceWidget
import br.com.zup.beagle.widget.core.Widget

data class FlexSingleWidget(
    val flex: Flex? = null,
    val child: Widget,
    override val appearance: Appearance? = null
) : AppearanceWidget

File: Sources/BeagleUI/Sources/Widgets/Layout/Flex/FlexSingleWidget.swift
//
//  Copyright © 2019 Daniel Tes. All rights reserved.
//

import UIKit

public struct FlexSingleWidget: Widget, HasAppearance {
    
    // MARK: - Public Properties
    
    public let appearance: Appearance?
    public let child: Widget
    public let flex: Flex
    
    // MARK: - Initialization
    
    public init(
        child: Widget,
        flex: Flex = Flex(),
        appearance: Appearance? = nil
    ) {
        self.child = child
        self.flex = flex
        self.appearance = appearance
    }
    
    // MARK: - Configuration
    
    public func apply(_ flex: Flex) -> FlexSingleWidget {
        return FlexSingleWidget(child: child, flex: flex)
    }
    
}

extension FlexSingleWidget: Renderable {
    public func toView(context: BeagleContext, dependencies: Renderable.Dependencies) -> UIView {
        let childView = child.toView(context: context, dependencies: dependencies)
        let view = UIView()
        view.addSubview(childView)
        view.applyAppearance(appearance)
        
        dependencies.flex.enableYoga(true, for: childView)
        dependencies.flex.setupFlex(flex, for: view)
        
        return view
    }
}
